eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientInstantiationCheck.java

```diff
@@ -0,0 +1,167 @@
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.CLASS_DEF,
+            TokenTypes.CTOR_DEF,
+            TokenTypes.METHOD_DEF,
+            TokenTypes.LITERAL_CLASS,
+            TokenTypes.OBJBLOCK
+        };
+    }
+
+    @Override
+    public void beginTree(DetailAST root) {
+        hasServiceClientAnnotation = true;
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        if (!hasServiceClientAnnotation) {
+            return;
+        }
+
+        switch (token.getType()) {
+            case TokenTypes.CLASS_DEF:
+                hasServiceClientAnnotation = hasServiceClientAnnotation(token);
+                break;
+            case TokenTypes.CTOR_DEF:
+                checkNonPublicOrProtectedConstructor(token);
+                break;
+            case TokenTypes.METHOD_DEF:
+                checkNonPublicStaticBuilderMethodName(token);
+                break;
+            case TokenTypes.LITERAL_CLASS:
+                checkServiceClientNaming(token);
+                break;
+            case TokenTypes.OBJBLOCK:
+                checkClassFieldFinal(token);
+                break;
+            default:
+                // Checkstyle complains if there's no default block in switch
+                break;
+        }
+    }
+
+    /**
+     *  Checks if the class is annotated with @ServiceClient annotation
+     *
+     * @param token the CLASS_DEF AST node
+     * @return true if the class is annotated with @ServiceClient, false otherwise.
+     */
+    private boolean hasServiceClientAnnotation(DetailAST token) {
+        DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+        if (modifiersToken != null) {
+            for (DetailAST ast = modifiersToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+                if (ast.getType() == TokenTypes.ANNOTATION) {
+                    DetailAST annotationIdent = ast.findFirstToken(TokenTypes.IDENT);
+                    if (annotationIdent != null && SERVICE_CLIENT.equals(annotationIdent.getText())) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     *  Checks for public or protected constructor for the service client class.
+     *  Log error if the service client has public or protected constructor.
+     *
+     * @param token the CTOR_DEF AST node
+     */
+    private void checkNonPublicOrProtectedConstructor(DetailAST token) {
+        DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+        AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+        if (accessModifier.equals(AccessModifier.PUBLIC) || accessModifier.equals(AccessModifier.PROTECTED)) {
+            log(modifiersToken, "Subclass of ServiceClient should not have any public or protected constructor.");
```

Comment:
It isn't a 'subclass of ServiceClient', it is a class annotated with `@ServiceClient`